sailfishusermanager.cpp

```diff
@@ -348,3 +371,97 @@ bool SailfishUserManager::hasAccessRights(uint uid_to_modify) {
 
     return true;
 }
+
+void SailfishUserManager::switchUser(uint uid)
+{
+    if (checkCallerUid() == SAILFISH_UNDEFINED_UID)
+        return;
+
+    if (m_switchUser) {
+        auto message = QStringLiteral("Already switching user");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::Failed, message);
+        return;
+    }
+
+    bool uidFound = false;
+    struct group *grent = getgrnam(USER_GROUP);
+    if (grent) {
+        for (int i = 0; grent->gr_mem[i]; i++) {
+            struct passwd *pw = getpwnam(grent->gr_mem[i]);
+            if (pw && pw->pw_uid == uid) {
+                uidFound = true;
+                break;
+            }
+        }
+    }
+    if (!uidFound) {
+        auto message = QStringLiteral("User not found");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::InvalidArgs, message);
+        return;
+    }
+
+    if (sd_seat_get_active("seat0", nullptr, &m_currentUid) < 0) {
+        auto message = QStringLiteral("Failed to get current user id");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::Failed, message);
+        return;
+    }
+
+    m_switchUser = uid;
+    m_timer->start();
+
+    if (!m_systemd)
+        m_systemd = new QDBusInterface(SYSTEMD_MANAGER_DEST, SYSTEMD_MANAGER_PATH, SYSTEMD_MANAGER_INTERFACE, QDBusConnection::systemBus());
+
+    // Stop user service
+    QDBusPendingCall call = m_systemd->asyncCall(SYSTEMD_MANAGER_STOP, USER_SERVICE.arg(m_currentUid), SYSTEMD_MANAGER_REPLACE);
+    QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(call, this);
+    connect(watcher, &QDBusPendingCallWatcher::finished, this, &SailfishUserManager::userServiceStop);
+}
+
+void SailfishUserManager::userServiceStop(QDBusPendingCallWatcher *reply)
+{
+    if (!reply->isValid()) {
+        auto message = QStringLiteral("Failed to stop current user session");
+        qCWarning(lcSUM) << message;
+        emit userSwitchFailed(m_switchUser);
+        m_switchUser = 0;
+        return;
+    }
+
+    // Stop autologin service
+    QDBusPendingCall call = m_systemd->asyncCall(SYSTEMD_MANAGER_STOP, AUTOLOGIN_SERVICE.arg(m_currentUid), SYSTEMD_MANAGER_REPLACE);
+    QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(call, this);
+    connect(watcher, &QDBusPendingCallWatcher::finished, this, &SailfishUserManager::autologinServiceStop);
+}
+
+void SailfishUserManager::autologinServiceStop(QDBusPendingCallWatcher *reply)
+{
+    if (!reply->isValid()) {
+        auto message = QStringLiteral("Failed to stop current user autologin");
+        qCWarning(lcSUM) << message;
+        emit userSwitchFailed(m_switchUser);
+        m_switchUser = 0;
+        return;
```

Comment:
I wonder if this should do something to recover. Like reboot.

Reply:
Adding old service restarts to attempt recovery